src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -18,159 +18,298 @@
-				case DELETE:
-					((SignChangeEvent) e).setLine(line, "");
-					break;
-				case SET:
-					assert delta != null;
-					((SignChangeEvent) e).setLine(line, (String) delta[0]);
-					break;
-			}
-		} else {
-			if (!sign.isOfType(b))
+		} else if (!lines) {
+			line = this.line.getOptionalSingle(event).orElse(1).intValue() - 1;
+			if (line < 0 || line > 3)
 				return;
-			final Sign s = (Sign) b.getState();
+		}
+		if (getTime() >= 0 && event instanceof SignChangeEvent && blocks.check(event, block -> block.equals(((SignChangeEvent) event).getBlock())) && !Delay.isDelayed(event)) {
+			String[] stringDelta = delta == null ? null : Arrays.copyOf(delta, delta.length, String[].class);
+			SignChangeEvent changeEvent = (SignChangeEvent) event;
 			switch (mode) {
+				case ADD:
+					assert stringDelta != null;
+					if (lines) {
+						List<String> list = Lists.newArrayList(changeEvent.getLines());
+						for (String string : stringDelta)
+							list.add(string);
+
+						stringDelta = list.toArray(new String[0]);
+					} else {
+						for (int i = 0; i < stringDelta.length; i++) {
+							String value = stringDelta.length > i ? (String) delta[i] : "";
+							stringDelta[i] = value + stringDelta[i];
+						}
+					}
+					//$FALL-THROUGH$
+				case REMOVE:
+				case REMOVE_ALL:
+					assert stringDelta != null;
+					stringDelta = ExprLore.handleRemove(StringUtils.join(stringDelta, "\n"), stringDelta[0], mode == ChangeMode.REMOVE_ALL).split("\n");
+					//$FALL-THROUGH$
 				case DELETE:
-					s.setLine(line, "");
-					break;
+					stringDelta = CollectionUtils.array("", "", "", "");
 				case SET:
-					assert delta != null;
-					s.setLine(line, (String) delta[0]);
+					// We need to ensure that it's clearing values without calling setLine twice.
+					if (mode == ChangeMode.SET) {
+						for (int i = 0; i < 4; i++)
+							stringDelta[i] = stringDelta.length > i ? (String) stringDelta[i] : "";
+					}
+					assert stringDelta != null;
+					if (lines) {
+						for (int i = 0; i < 4; i++) {
+							String value = stringDelta.length > i ? (String) stringDelta[i] : "";
+							if (serializer != null) {
+								changeEvent.line(i, serializer.deserialize(BungeeConverter.convert(ChatMessages.parseToArray(value))));
+								continue;
+							}
+							changeEvent.setLine(i, value);
+						}
+						break;
+					}
+					changeEvent.setLine(line, (String) delta[0]);
+					break;
+				default:
 					break;
 			}
-			if (hasUpdateBooleanBoolean) {
-				try {
-					s.update(false, false);
-				} catch (final NoSuchMethodError err) {
-					hasUpdateBooleanBoolean = false;
-					s.update();
-				}
-			} else {
-				s.update();
-			}
+			return;
 		}
+		int finalLine = line;
+		blocks.stream(event)
+				.map(Block::getState)
+				.filter(Sign.class::isInstance)
+				.map(Sign.class::cast)
+				.forEach(sign -> {
+					String[] stringDelta = delta == null ? null : Arrays.copyOf(delta, delta.length, String[].class);
+					switch (mode) {
```

Comment:
All the same issues from above apply here.
Is there any way to unduplicate this code, too?